In [1]:
# loading libraries
import numpy as np
import pandas as pd

<hr>

## Question 1

If you're given two lists:

A = [1, 2, 3, 4, 5, 6], B = [13, 21, 34]

The task is to create a list with the elements of A and B in a single dimension with output:

A_B = [1, 2, 3, 4, 5, 6, 13, 21, 34]

Which of the following option is the best way to create this list?

In [2]:
A_extend = [1, 2, 3, 4, 5, 6]
A_append = [1, 2, 3, 4, 5, 6]
B = [13, 21, 34]

A_extend.extend(B)
A_append.append(B)

print("Extend:", A_extend)
print("Append:", A_append)

Extend: [1, 2, 3, 4, 5, 6, 13, 21, 34]
Append: [1, 2, 3, 4, 5, 6, [13, 21, 34]]


#### <font color = 'red'>Question 1 ANSWER:</font>
**A.extend(B)**

<hr>

## Question 2

How do you create an identity matrix in Python?

In [3]:
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [4]:
np.array([1, 0, 0], [0, 1, 0], [0, 0, 1])

ValueError: only 2 non-keyword arguments accepted

In [5]:
np.array[(1, 0, 0), (0, 1, 0), (0, 0, 1)]

TypeError: 'builtin_function_or_method' object is not subscriptable

In [6]:
eye(3)

NameError: name 'eye' is not defined

#### <font color = 'red'>Question 2 ANSWER:</font>
**np.identity(3)**

<hr>

## Question 3

Which of the following fuel type code has the lowest average fuel cost per unit burned?

In [7]:
# load data set
fuel = pd.read_csv('fuel_ferc1.csv')

# preview a sample
fuel.sample(3)

record_id  utility_id_ferc1  report_year  \
9878   f1_fuel_2002_12_164_0_7               164         2002   
22242   f1_fuel_2012_12_45_1_7                45         2012   
8722   f1_fuel_2001_12_127_0_4               127         2001   

      plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
9878          knox lee                 gas       mcf        7032985.0   
22242        cliffside                coal       ton         734801.0   
8722             gavin                coal       ton       73761030.0   

       fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
9878                 1.031                      3.492   
22242               24.538                    103.880   
8722                23.146                     39.546   

       fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
9878                          3.492                3.390  
22242                       102.880                4.141  
8722                         35.488                1.709

In [8]:
# group by 'fuel_type_code_pudl' and calculate the means of 'fuel_cost_per_unit_burned' per group
fuel.groupby('fuel_type_code_pudl').fuel_cost_per_unit_burned.mean().sort_values()

fuel_type_code_pudl
gas          13.659397
other        18.253856
waste        19.518122
coal         67.421830
oil         168.877086
nuclear    4955.157002
Name: fuel_cost_per_unit_burned, dtype: float64

#### <font color = 'red'>Question 3 ANSWER:</font>
**Gas**

<hr>

## Question 4

What is the standard deviation and 75th percentile of the measure of energy per unit(fuel_mmbtu_per_unit) in two decimal places?

In [9]:
print("Standard deviation:", round(fuel.fuel_mmbtu_per_unit.std(), 2))
print("75th percentile:", round(fuel.fuel_mmbtu_per_unit.quantile(0.75), 2))

Standard deviation: 10.6
75th percentile: 17.01


#### <font color = 'red'>Question 4 ANSWER:</font>
**10.60 and 17.01**

<hr>

## Question 5

What is the skewness and kurtosis for the fuel quantity burned in two decimal places?

In [10]:
print("Skewness:", round(fuel.fuel_qty_burned.skew(), 2))
print("Kurtosis:", round(fuel.fuel_qty_burned.kurt(), 2))

Skewness: 15.85
Kurtosis: 651.37


#### <font color = 'red'>Question 5 ANSWER:</font>
**15.85 and 651.37**

<hr>

## Question 6

Which feature has missing values and what is the total number of missing values and percentage of the missing rows as a factor of the total number of rows in three decimal places?<br>
(Enter answer in the format: Feature: xxx, Total: xxx, Percent: xxx)

In [11]:
# get total number of missing values per column
fuel.isna().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [12]:
# percentage of rows with missing values
total_missing = fuel.fuel_unit.isna().sum()
total_rows = fuel.shape[0]
percentage_missing = (total_missing / total_rows) * 100
print("Percentage of missing rows:", round(percentage_missing, 3))

Percentage of missing rows: 0.61


#### <font color = 'red'>Question 6 ANSWER:</font>
**Feature: fuel_unit, Total: 180, Percent: 0.610**

<hr>

## Question 7

The feature with missing values falls under what category? What missing value imputation technique would you use?

In [13]:
# getting the data type of fuel_unit
fuel.fuel_unit.dtypes

dtype('O')

Data type 'O' stands for 'object', which is a categorical variable and, therefore, cannot not have mean or median values.

#### <font color = 'red'>Question 7 ANSWER:</font>
**Categorical and mode imputation**

<hr>

## Question 8

Which of the features has the second and third lowest correlation with the Fuel Cost Per Unit Burned?

In [14]:
corr = fuel.corr()
corr.fuel_cost_per_unit_burned.sort_values()

utility_id_ferc1               -0.037863
fuel_qty_burned                -0.018535
fuel_mmbtu_per_unit            -0.010034
fuel_cost_per_mmbtu            -0.000437
fuel_cost_per_unit_delivered    0.011007
report_year                     0.013599
fuel_cost_per_unit_burned       1.000000
Name: fuel_cost_per_unit_burned, dtype: float64

#### <font color = 'red'>Question 8 ANSWER:</font>
**fuel_qty_burned and fuel_mmbtu_per_unit**

<hr>

## Question 9

For the fuel type coal, what is the percentage change in the fuel cost per unit burned in 1998 compared to 1994?

In [15]:
# select records with fuel type coal for 1994 and 1998
coal_94 = fuel.loc[(fuel['fuel_type_code_pudl'] == 'coal') & ((fuel['report_year'] == 1994))]
coal_98 = fuel.loc[(fuel['fuel_type_code_pudl'] == 'coal') & ((fuel['report_year'] == 1998))]

coal_94_sum = coal_94.fuel_cost_per_unit_burned.sum()
coal_98_sum = coal_98.fuel_cost_per_unit_burned.sum()

pct_change = (coal_98_sum - coal_94_sum) / coal_94_sum * 100
print("Percentage change:", round(pct_change))

Percentage change: -21.0


#### <font color = 'red'>Question 9 ANSWER:</font>
**-21%**

<hr>

## Question 10

Which year has the highest average fuel cost per unit delivered?

In [16]:
# group by 'report_year' and then calculate the mean of 'fuel_cost_per_unit_burned' per year
fuel.groupby('report_year').fuel_cost_per_unit_delivered.mean().sort_values(ascending = False)[:3]

report_year
1997    11140.197239
1996     9196.705948
2000      985.362877
Name: fuel_cost_per_unit_delivered, dtype: float64

#### <font color = 'red'>Question 10 ANSWER:</font>
**1997**